pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+> Analysis of JSON Schema](schema-static-analysis) article by [Greg
+> Dennis](https://www.linkedin.com/in/gregdennis/).
+
+However, there is a small set of keywords whose evaluation depend on the
+*scope* they are in. These keywords are `$ref`, `$dynamicRef`,
+`unevaluatedItems`, and `unevaluatedProperties`. Additionally, there is also a
+set of keywords that *affect* the scope they are declared in. These keywords
+are `$id`, `$schema`, `$anchor`, and `$dynamicAnchor`.
+
+JSON Schema defines two types of scopes for the purpose of URI resolution: the
+*lexical scope* and the *dynamic scope*. Understanding how these scopes work is
+essential for mastering some of the most advanced (and often confusing!)
+features of JSON Schema, such as dynamic referencing.
+
+Schema Resources
+----------------
+
+Before we jump into lexical and dynamic scopes, lets review some JSON Schema
+fundamentals.
+
+The `$id` keyword defines the URI of a
+[schema](https://json-schema.org/learn/glossary#schema). While this keyword is
+typically set at the top level, any
+[subschema](https://json-schema.org/learn/glossary#subschema) may declare it to
+distinguish itself with a different URI. For example, the following schema
+defines 4 identifiers, some relative and some absolute:
+
+![A JSON Schema with multiple identifiers](/img/posts/2024/understanding-lexical-dynamic-scopes/scopes-multiple-ids.webp)
+
+[In JSON Schema
+parlance](https://json-schema.org/draft/2020-12/json-schema-core#name-root-schema-and-subschemas-),
+we say that the `$id` keyword introduces a new *schema resource*, and that the
+top level schema resource is referred to as the *root schema resource*.
+
+Consider the following example. This schema consists of 3 schema resources,
+each highlighted using a different color: the root schema resource (red), the
+schema resource at `/properties/foo` (blue), and the schema resource at
+`/properties/bar` (green). Note that the subschema at `/properties/baz` is part
+of the root schema resource, as it does not introduce a new identifier:
+
+![A JSON Schema that consists of 3 schema resources](/img/posts/2024/understanding-lexical-dynamic-scopes/scopes-schema-resources.webp)
+
+Note that children schema resources are not considered part of the parent
+schema resource. For example, in the previous figure, `https://example.com/foo`
+or `https://example.com/bar` are *separate* schema resources and not part of
+the root schema resource, despite their structural relationship.
+
+Schemas as Directed Graphs
+--------------------------
+
+JSON Schema is a recursive data structure. In the context of schema resources,
+this means that a schema resource may introduce nested schema resources (like
+we saw on the previous section) and use referencing keywords (like `$ref`) to
+point to external schema resources, effectively creating a directed graph of
+schema resources. 
+
+Consider the following example. In the top left, a root schema resource named
```

Comment:
However, we do use "in" when there are multiple directions, like "top right" or "bottom left".

English is dumb.